### PR TITLE
rust/applayer - derive utility functions for app-layer event enums - v7

### PR DESCRIPTION
**Pull request by jasonish**
This PR shows how we can use a Rust procedural macro to derive AppLayer event functions from an emum.

Given an enum like:
```
#[derive(Debug, PartialEq, AppLayerEvent)]
pub enum DNSEvent {
    MalformedData,
    NotRequest,
    NotResponse,
    ZFlagSet,
}
```

supporting code will be added allowing the protocol to register `get_event_info` and `get_event_info_by_id` callbacks like:
```
get_eventinfo: Some(DNSEvent::get_event_info),
get_eventinfo_byid: Some(DNSEvent::get_event_info_by_id),
```

All the parser needs to do is enumerate the different events, the macro will take care of generating the support code.

Redmine issue:
https://redmine.openinfosecfoundation.org/issues/4154

Previous PR:
https://github.com/OISF/suricata/pull/5501

Changes from last PR:
- Convert all Rust parsers.

### CODE REVIEW

File: rust/derive/src/applayerevent.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            }
+
+        }
+    };
+
+    proc_macro::TokenStream::from(expanded)
+}
+
+/// Transform names such as "OneTwoThree" to "one_two_three".
+pub fn transform_name(in_name: &str) -> String {
```
Comment (dbcfd) [inline rust/derive/src/applayerevent.rs:97]: Does it make sense to use something like https://crates.io/crates/Inflector here?

File: rust/derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use proc_macro::TokenStream;
+
+mod applayerevent;
+
+/// The `AppLayerEvent` derive macro generates a `AppLayerEvent` trait
+/// implementation for enums that define AppLayerEvents.
+///
+/// Example usage (DNS app-layer events):
+///
+/// #[derive(AppLayerEvent)]
```
Comment (dbcfd) [inline rust/derive/src/lib.rs:29]: Maybe add a rust/derive/tests/test.rs to demonstrate and test.

File: rust/derive/src/applayerevent.rs (hunk 1 above)
Comment (jasonish) [reply rust/derive/src/applayerevent.rs:97]: Not too big of a fan of pulling in a dependency to solve something that can be done pretty easily in a function or 2.

File: rust/derive/src/lib.rs (hunk 2 above)
Comment (jasonish) [reply rust/derive/src/lib.rs:29]: Yeah, docs/examples need a little more detail.

File: rust/derive/Cargo.toml.in [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "suricata-derive"
+version = "@PACKAGE_VERSION@"
+edition = "2018"
+
+[lib]
+proc-macro = true
+path = "@e_rustdir@/derive/src/lib.rs"
+
+[dependencies]
```
Comment (victorjulien) [inline rust/derive/Cargo.toml.in:10]: are these build dependencies only? Or do they end up in the compiled bin?

Comment (jasonish) [reply rust/derive/Cargo.toml.in:10]: They are build only, but can't be labeled as [build-dependencies].  Within the scope of the derive crate, they are runtime dependencies, but only used by Suricata at a build time.

However, it doesn't appear the derive crate can be labeled as a build-dependency either, however, it does not end up in the output binary. Instead the derive crate is compiled and loaded by the compiler as a "compiler" plugin.  This becomes more visible in a cross compile scenario. The derive plugin would be compiled for the host environment, not the target environment.

Even though they already show up in our dependency tree, nothing triggers their usage yet so they will be new crates that we vendor.  They're rather ubiquitous for anyone building procedural macros.

File: rust/derive/src/applayerevent.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            for (i, v) in (&data.variants).into_iter().enumerate() {
+                fields.push(v.ident.clone());
+                let name = transform_name(&v.ident.to_string());
+                let cname = format!("{}\0", name);
+                names.push(name);
+                cstrings.push(cname);
+                vals.push(i as i32);
+            }
+        }
+        _ => unimplemented!(),
```
[the remark this answers is not in the document]
Comment (jasonish) [reply rust/derive/src/applayerevent.rs:43]: Fixed with a panic message:
```
error: proc-macro derive panicked
   --> src/dns/dns.rs:140:10
    |
140 | #[derive(AppLayerEvent)]
    |          ^^^^^^^^^^^^^
    |
    = help: message: AppLayerEvent can only be derived for enums

error: aborting due to previous error

error: could not compile `suricata`

To learn more, run the command again with --verbose.
```
I also looked at the compile_error!() macro, but it doesn't provide the line where the macro is used incorrectly.

File: rust/derive/Cargo.toml.in (hunk 3 above)
Comment (victorjulien) [reply rust/derive/Cargo.toml.in:10]: The last sentence on whether they will show up in the vendored crates is a bit unclear to me. Will they be included, but not end up in the compiled bin?

Comment (jasonish) [reply rust/derive/Cargo.toml.in:10]: They will be included in our vendor directory, compiled and run, but not in the output binary. However, their output (the generated code) will be in our output binary if that makes sense.

Much like the output of yacc, but where yacc is compiled and run as part of the build.

**Comment by jasonish** [comment]
Updated version at https://github.com/OISF/suricata/pull/6318.

**Pull request closed without merging**